host/share/man/man1/quilt.1
.\" Created by Martin Quinson from the tex documentation
.\"
.TH quilt 1 "Dec 17, 2013" "quilt"

.SH NAME
quilt \- tool to manage series of patches

.SH SYNOPSIS
.B quilt
[-h] command [options]

.SH DESCRIPTION
Quilt is a tool to manage large sets of patches by keeping track of the
changes each patch makes. Patches can be applied, un-applied, refreshed,
etc. The key philosophical concept is that your primary output is patches.

With quilt, all work occurs within a single directory tree. Commands can be
invoked from anywhere within the source tree. They are of the form
.B quilt cmd
similar to CVS, svn or git commands. They can be abbreviated as long as the specified
part of the command is unique. All commands print some help text with
.B quilt cmd -h.

Quilt manages a stack of patches. Patches are applied incrementally on top
of the base tree plus all preceding patches. They can be pushed on top of
the stack
.RB ( "quilt push" ),
and popped off the stack
.RB ( "quilt pop" ).
Commands are available for querying the contents of the series file
.RB ( "quilt series" ,
see below), the contents of the stack
.RB ( "quilt applied" , " quilt previous" , " quilt top" ),
and the patches that are not applied at a particular moment
.RB ( "quilt next" , " quilt unapplied" ).
By default, most commands apply to the topmost patch on the stack.

Patch files are located in the
.I patches
sub-directory of the source tree (see EXAMPLE OF WORKING TREE below). The
.I QUILT_PATCHES
environment variable can be used to override this location. When not
found in the current directory, that subdirectory is searched
recursively in the parent directories (this is similar to the way
.I git
searches for its configuration files). The
.I patches
directory may contain sub-directories. It may also be a symbolic link
instead of a directory.

A file called
.I series
contains a list of patch file names that defines the order in which patches
are applied. Unless there are means by which series files can be generated
automatically, it is usually provided along with a set of patches. In this
file, each patch file name is on a separate line. Patch files are identified
by path names that are relative to the
.I patches
directory; patches may be in sub-directories below this directory. Lines
in the series file that start with a hash character (#) are ignored.
Patch options, such as the strip level or whether the patch is reversed, can
be added after each patch file name. Options are introduced by a space,
separated by spaces, and follow the syntax of the patch(1) options
(e.g. -p2). Quilt records patch options automatically when a command
supporting them is used. Without options, strip level 1 is assumed.
You can also add a comment after each patch file name and options, introduced
by a space followed by a hash character. When
quilt adds, removes, or renames patches, it automatically updates the series
file. Users of quilt can modify series files while some patches are
applied, as long as the applied patches remain in their original order.

Different series files can be used to assemble patches in different ways,
corresponding for example to different development branches.

Before a patch is applied (or ``pushed on the stack''), copies of all files
the patch modifies are saved to the
.RI .pc/ patch
directory. The patch is added to the list of currently applied patches
(.pc/applied-patches). Later when a patch is regenerated
.RB ( "quilt refresh" ),
the backup copies in
.RI .pc/ patch
are compared with the current versions of the files in the source tree using
GNU diff.

Documentation related to a patch can be put at the beginning of a patch
file.  Quilt is careful to preserve all text that precedes the actual patch
when doing a refresh. (This is limited to patches in unified format; see
.B diff
documentation).

The series file is looked up in the .pc directory, in the root of the source
tree, and in the patches directory.  The first series file that is found is
used. This may also be a symbolic link, or a file with multiple hard links.
Usually, only one series file is used for a set of patches, so the
patches sub-directory is a convenient location.

The .pc directory and its sub-directories cannot be relocated, but it can be
a symbolic link. While patches are applied to the source tree, this
directory is essential for many operations, including taking patches off the
stack
.RB ( "quilt pop" ),
and refreshing patches
.RB ( "quilt refresh" ).
Files in the .pc directory are automatically removed when they are
no longer needed, so there is no need to clean up manually.

.SH QUILT COMMANDS REFERENCE


.IP "\fBadd\fP [-P patch] {file} ... " 4

Add one or more files to the topmost or named patch.  Files must be
added to the patch before being modified.  Files that are modified by
patches already applied on top of the specified patch cannot be added.

.IP "    -P patch" 8

Patch to add files to.


.IP "\fBannotate\fP [-P patch] {file} " 4

Print an annotated listing of the specified file showing which
patches modify which lines. Only applied patches are included.

.IP "    -P patch" 8

Stop checking for changes at the specified rather than the
topmost patch.


.IP "\fBapplied\fP [patch] " 4

Print a list of applied patches, or all patches up to and including the
specified patch in the file series.


.IP "\fBdelete\fP [-r] [--backup] [patch|-n] " 4

Remove the specified or topmost patch from the series file.  If the
patch is applied, quilt will attempt to remove it first. (Only the
topmost patch can be removed right now.)

.IP "    -n" 8
Delete the next patch after topmost, rather than the specified
or topmost patch.

.IP "    -r" 8
Remove the deleted patch file from the patches directory as well.

.IP "    --backup" 8

Rename the patch file to patch~ rather than deleting it.
Ignored if not used with `-r'.


.IP "\fBdiff\fP [-p n|-p ab] [-u|-U num|-c|-C num] [--combine patch|-z] [-R] [-P patch] [--snapshot] [--diff=utility] [--no-timestamps] [--no-index] [--sort] [--color[=always|auto|never]] [file ...] " 4

Produces a diff of the specified file(s) in the topmost or specified
patch.  If no files are specified, all files that are modified are
included.

.IP "    -p n" 8
Create a -p n style patch (-p0 or -p1 are supported).

.IP "    -p ab" 8
Create a -p1 style patch, but use a/file and b/file as the
original and new filenames instead of the default
dir.orig/file and dir/file names.

.IP "    -u, -U num, -c, -C num" 8

Create a unified diff (-u, -U) with num lines of context. Create
a context diff (-c, -C) with num lines of context. The number of
context lines defaults to 3.

.IP "    --no-timestamps" 8

Do not include file timestamps in patch headers.

.IP "    --no-index" 8

Do not output Index: lines.

.IP "    -z" 8
Write to standard output the changes that have been made
relative to the topmost or specified patch.

.IP "    -R" 8
Create a reverse diff.

.IP "    -P patch" 8

Create a diff for the specified patch.  (Defaults to the topmost
patch.)

.IP "    --combine patch" 8

Create a combined diff for all patches between this patch and
the patch specified with -P. A patch name of `-' is equivalent
to specifying the first applied patch.

.IP "    --snapshot" 8

Diff against snapshot (see `quilt snapshot -h').

.IP "    --diff=utility" 8

Use the specified utility for generating the diff. The utility
is invoked with the original and new file name as arguments.

.IP "    --color[=always|auto|never]" 8

Use syntax coloring (auto activates it only if the output is a tty).

.IP "    --sort" 8
Sort files by their name instead of preserving the original order.


.IP "\fBedit\fP file ... " 4

Edit the specified file(s) in $EDITOR after adding it (them) to
the topmost patch.


.IP "\fBfiles\fP [-v] [-a] [-l] [--combine patch] [patch] " 4

Print the list of files that the topmost or specified patch changes.

.IP "    -a" 8
List all files in all applied patches.

.IP "    -l" 8
Add patch name to output.

.IP "    -v" 8
Verbose, more user friendly output.

.IP "    --combine patch" 8

Create a listing for all patches between this patch and
the topmost or specified patch. A patch name of `-' is
equivalent to specifying the first applied patch.



.IP "\fBfold\fP [-R] [-q] [-f] [-p strip-level] " 4

Integrate the patch read from standard input into the topmost patch:
After making sure that all files modified are part of the topmost
patch, the patch is applied with the specified strip level (which
defaults to 1).

.IP "    -R" 8
Apply patch in reverse.

.IP "    -q" 8
Quiet operation.

.IP "    -f" 8
Force apply, even if the patch has rejects. Unless in quiet mode,
apply the patch interactively: the patch utility may ask questions.

.IP "    -p strip-level" 8

The number of pathname components to strip from file names
when applying patchfile.


.IP "\fBfork\fP [new_name] " 4

Fork the topmost patch.  Forking a patch means creating a verbatim copy
of it under a new name, and use that new name instead of the original
one in the current series.  This is useful when a patch has to be
modified, but the original version of it should be preserved, e.g.
because it is used in another series, or for the history.  A typical
sequence of commands would be: fork, edit, refresh.

If new_name is missing, the name of the forked patch will be the current
patch name, followed by `-2'.  If the patch name already ends in a
dash-and-number, the number is further incremented (e.g., patch.diff,
patch-2.diff, patch-3.diff).


.IP "\fBgraph\fP [--all] [--reduce] [--lines[=num]] [--edge-labels=files] [-T ps] [patch] " 4

Generate a dot(1) directed graph showing the dependencies between
applied patches. A patch depends on another patch if both touch the same
file or, with the --lines option, if their modifications overlap. Unless
otherwise specified, the graph includes all patches that the topmost
patch depends on.
When a patch name is specified, instead of the topmost patch, create a
graph for the specified patch. The graph will include all other patches
that this patch depends on, as well as all patches that depend on this
patch.

.IP "    --all" 8
Generate a graph including all applied patches and their
dependencies. (Unapplied patches are not included.)

.IP "    --reduce" 8

Eliminate transitive edges from the graph.

.IP "    --lines[=num]" 8

Compute dependencies by looking at the lines the patches modify.
Unless a different num is specified, two lines of context are
included.

.IP "    --edge-labels=files" 8

Label graph edges with the file names that the adjacent patches
modify.

.IP "    -T ps" 8
Directly produce a PostScript output file.


.IP "\fBgrep\fP [-h|options] {pattern} " 4

Grep through the source files, recursively, skipping patches and quilt
meta-information. If no filename argument is given, the whole source
tree is searched. Please see the grep(1) manual page for options.

.IP "    -h" 8
Print this help. The grep -h option can be passed after a
double-dash (--). Search expressions that start with a dash
can be passed after a second double-dash (-- --).


.IP "\fBheader\fP [-a|-r|-e] [--backup] [--strip-diffstat] [--strip-trailing-whitespace] [patch] " 4

Print or change the header of the topmost or specified patch.

.IP "    -a, -r, -e" 8

Append to (-a) or replace (-r) the exiting patch header, or
edit (-e) the header in $EDITOR. If none of these options is
given, print the patch header.

.IP "    --strip-diffstat" 8

Strip diffstat output from the header.

.IP "    --strip-trailing-whitespace" 8

Strip trailing whitespace at the end of lines of the header.

.IP "    --backup" 8

Create a backup copy of the old version of a patch as patch~.


.IP "\fBimport\fP [-p num] [-R] [-P patch] [-f] [-d {o|a|n}] patchfile ... " 4

Import external patches.  The patches will be inserted following the
current top patch, and must be pushed after import to apply them.

.IP "    -p num" 8

Number of directory levels to strip when applying (default=1)

.IP "    -R" 8

Apply patch in reverse.

.IP "    -P patch" 8

Patch filename to use inside quilt. This option can only be
used when importing a single patch.

.IP "    -f" 8
Overwrite/update existing patches.

.IP "    -d {o|a|n}" 8

When overwriting in existing patch, keep the old (o), all (a), or
new (n) patch header. If both patches include headers, this option
must be specified. This option is only effective when -f is used.


.IP "\fBmail\fP {--mbox file|--send} [-m text] [-M file] [--prefix prefix] [--sender ...] [--from ...] [--to ...] [--cc ...] [--bcc ...] [--subject ...] [--reply-to message] [--charset ...] [--signature file] [first_patch [last_patch]] " 4

Create mail messages from a specified range of patches, or all patches in
the series file, and either store them in a mailbox file, or send them
immediately. The editor is opened with a template for the introduction.
Please see /workdir/openwrt/staging_dir/host/share/doc/quilt/README.MAIL for details.
When specifying a range of patches, a first patch name of `-' denotes the
first, and a last patch name of `-' denotes the last patch in the series.

.IP "    -m text" 8

Text to use as the text in the introduction. When this option is
used, the editor will not be invoked, and the patches will be
processed immediately.

.IP "    -M file" 8

Like the -m option, but read the introduction from file.

.IP "    --prefix prefix" 8

Use an alternate prefix in the bracketed part of the subjects
generated. Defaults to `patch'.

.IP "    --mbox file" 8

Store all messages in the specified file in mbox format. The mbox
can later be sent using formail, for example.

.IP "    --send" 8

Send the messages directly.

.IP "    --sender" 8

The envelope sender address to use. The address must be of the form
`user@domain.name'. No display name is allowed.

.IP "    --from, --subject" 8

The values for the From and Subject headers to use. If no --from
option is given, the value of the --sender option is used.

.IP "    --to, --cc, --bcc" 8

Append a recipient to the To, Cc, or Bcc header.

.IP "    --charset" 8

Specify a particular message encoding on systems which don't use
UTF-8 or ISO-8859-15. This character encoding must match the one
used in the patches.

.IP "    --signature file" 8

Append the specified signature to messages (defaults to ~/.signature
if found; use `-' for no signature).

.IP "    --reply-to message" 8

Add the appropriate headers to reply to the specified message.


.IP "\fBnew\fP [-p n|-p ab] {patchname} " 4

Create a new patch with the specified file name, and insert it after the
topmost patch. The name can be prefixed with a sub-directory name, allowing
for grouping related patches together.

.IP "    -p n" 8
Create a -p n style patch (-p0 or -p1 are supported).

.IP "    -p ab" 8
Create a -p1 style patch, but use a/file and b/file as the
original and new filenames instead of the default
dir.orig/file and dir/file names.


host/share/emacs/site-lisp/quilt.el
;;; quilt.el --- a minor mode for working with files in quilt

;; Copyright 2005-2007  Matt Mackall <mpm@selenic.com>
;;  Satoru Takeuchi <satoru.takeuchi@gmail.com> took over this package
;;  from Matt Mackall.

;; Author: Saotru takeuchi <satoru.takeuchi@gmail.com>
;;
;; This software may be used and distributed according to the terms
;; of the GNU General Public License, incorporated herein by reference.

;;; Commentary:
;;

;;  Add (load "~/quilt.el") to your .emacs file

;;; History:
;;

;;; Code:

(defun quilt-buffer-file-name-safe ()
  "Return buffer file name.  If buffer is not associated with any file, return nil."
  (let ((fn buffer-file-name))
    (if (and fn (file-exists-p fn))
	fn)))

(defun quilt-bottom-p ()
  "Return t if there is on the bottom of patch stack, return nil if otherwise."
  (if (> (call-process "quilt" nil nil nil "applied") 0) 1))

(defun quilt--get-config-variable (var)
  "Return the value of a configuration variable. Return nil if it is unset."
  (or (with-current-buffer (generate-new-buffer " *cmd")
        (shell-command
         (concat "if [ -f ~/.quiltrc ]; then"
                 "  . ~/.quiltrc ;"
                 "elif [ -f /etc/quilt.quiltrc ]; then"
                 "  . /etc/quilt.quiltrc ;"
                 "fi ;"
                 "echo -n $" var)
         t)
        (unwind-protect
            (let ((v (buffer-string)))
              (if (string= "" (buffer-string))
                  nil
                v))
          (kill-buffer (current-buffer))))
      (getenv var)))

(defun quilt--per-project-patches-directory ()
  (let ((qd (quilt-dir)))
    (if qd
        (let ((project-config (concat qd "/"
                                      (quilt-pc-directory) "/.quilt_patches")))
          (if (file-readable-p project-config)
              (with-temp-buffer
                (insert-file-contents-literally project-config)
                (substring (buffer-string) 0 -1)))))))

(defun quilt-patches-directory ()
  "Return the location of patch files."
  (or (quilt--per-project-patches-directory)
      (quilt--get-config-variable "QUILT_PATCHES")
      "patches"))

(defun quilt-pc-directory ()
  "Return the location of the quilt working state directory."
  (or (quilt--get-config-variable "QUILT_PC")
      ".pc"))

(defun quilt-patches-prefix ()
  "Return the value of the QUILT_PATCHES_PREFIX config variable. Return nil if it is unset."
  (quilt--get-config-variable "QUILT_PATCHES_PREFIX"))

(defun quilt-find-dir (fn &optional prefn)
  "Return the top level dir of quilt from FN."
  (if (or (not fn) (equal fn "/") (equal fn prefn))
      nil
    (let ((d (file-name-directory fn)))
      (if (file-accessible-directory-p (concat d "/" (quilt-pc-directory)))
	  d
	(quilt-find-dir (directory-file-name d) d)))))

(defun quilt-dir (&optional fn)
  "Return the top level dir of quilt from FN.  FN is just a hint and find from other way if FN is nil."
  (quilt-find-dir (if fn fn
		    (let ((fn2 (quilt-buffer-file-name-safe)))
		      (if fn2 fn2
			(if default-directory
			    (expand-file-name default-directory)))))))

(defun quilt-drop-dir (fn)
  "Return the relative path of FN based on quilt top directory."
  (let ((d (quilt-find-dir fn)))
    (substring fn (length d) (length fn))))

(defun quilt-p (&optional fn)
  "Check if FN is in a quilt tree."
  (if (quilt-dir fn) 't nil))

(defun quilt-save ()
  "Save all buffers associated with current quilt tree."
  (save-some-buffers nil 'quilt-p))

(defun quilt-owned-p (fn)
  "Check if FN is a file which quilt handles."
  (if (not fn)
      nil
    (let ((pd (file-name-nondirectory
		(directory-file-name (file-name-directory fn)))))
      (and
       (not (string-match "\\(~$\\|\\.rej$\\)" fn))
       (not (equal pd (quilt-patches-directory)))
       (not (equal pd (quilt-pc-directory)))
       (quilt-p fn)))))

(defun quilt-cmd (cmd &optional buf)
  "Execute CMD, a quilt subcommand, at the top of quilt tree associated with BUF."
  (let ((d default-directory)
	 (qd (quilt-dir)))
    (if (not qd)
	(shell-command (concat "quilt " cmd) buf)
      (cd qd)
      (unwind-protect ; make sure to cd back even if an erro occurs.
	  (shell-command (concat "quilt " cmd) buf)
	(cd d)))))

(defun quilt-cmd-to-string (cmd)
  "Execute CMD, a quilt subcommand, at the top of quilt tree associated with BUF and return its output string."
  (let ((d default-directory)
	 (qd (quilt-dir)))
    (if (not qd)
	nil
      (cd qd)
      (unwind-protect ; make sure to cd back even if an error occurs.
	  (shell-command-to-string (concat "quilt " cmd))
	(cd d)))))

(defun quilt-cmd-to-list (cmd)
  "Execute CMD, a quilt sumcommand at the top of quilt tree associated with BUF.."
  "Return the lines of the command output as elements of a list."
  (let ((s (quilt-cmd-to-string cmd)))
    (if s
        (split-string s "\n" t))))

(defun quilt-applied-list ()
  "Return the list of the applied patch names."
  (quilt-cmd-to-list "applied"))

(defun quilt-file-list ()
  "Return the list of the file names associated with current patch."
  (quilt-cmd-to-list "files"))

(defun quilt-patch-list ()
  "Return the list of the name of patches."
  (quilt-cmd-to-list "series"))

(defun quilt-files-affected (&optional first last)
  "Return the file names which modified from FIRST to LAST."
  "Omitted args are considered as current patch."
  (let ((qd (quilt-dir))
        files fp)
    (when qd
      (setq files (quilt-cmd-to-list
                   (if last
                       (if (equal first last)
                           (concat "files " first)
                         (concat "files --combine " first last))
                     (if first
                         (concat "files --combine " first)
                       "files"))))
      (setq fp files)
      (while fp
        (setcar fp (concat qd (car fp)))
        (setq fp (cdr fp)))
      files)))

(defun quilt-top-patch ()
  "Return the top patch name.  return nil if there is the bottom of patch stack."
  (if (quilt-bottom-p)
      nil
    (substring (quilt-cmd-to-string "top")  0 -1)))

(defun quilt-complete-list (p l)
  "Call 'completing-read' with prompt P and list L."
"Convert L to an alist using indices as keys.  Note that this function modifies L."
  (let ((list l)
	(n 0))
    (while list
      (setcar list (cons (car list) n))
      (setq list (cdr list))
      (setq n (1+ n))))
  (completing-read p l nil t))

(defun quilt--strip-patchname (pn)
  "Return the name of patch PN sans the path to the patches directory."
  (let ((patches-path (concat (quilt-patches-directory) "/")))
    (substring pn (length patches-path))))

(defvar quilt-edit-top-only 't)
(defun quilt-editable (f)
  "Return t if F is editable in terms of current patch.  Return nil if otherwise."
  (let ((qd (quilt-dir))
	 (fn (quilt-drop-dir f))
	 dirs result)
    (if qd
	(if (quilt-bottom-p)
	    (quilt-cmd "applied")	; to print error message
	  (setq dirs (if quilt-edit-top-only
			 (list (let ((patch (substring (quilt-cmd-to-string "top")  0 -1)))
				 (if (quilt-patches-prefix)
				     (quilt--strip-patchname patch)
				   patch)))
			 (cdr (cdr (directory-files (concat qd (quilt-pc-directory) "/"))))))
	  (while (and (not result) dirs)
	    (if (file-exists-p (concat qd (quilt-pc-directory) "/" (car dirs) "/" fn))
		(setq result t)
	      (setq dirs (cdr dirs))))
	  result))))

(defun quilt-short-patchname ()
  "Return shortened name of top patch.  Return \"none\" when on the bottom patch stack."
  (let ((p (quilt-top-patch)))
    (if (not p)
	"none"
      (let ((p2 (file-name-sans-extension (file-name-nondirectory p))))
	   (if (< (length p2) 10)
	       p2
	     (concat (substring p2 0 8) ".."))))))

(defvar quilt-mode-line nil)
(make-variable-buffer-local 'quilt-mode-line)

(defun quilt-update-modeline ()
  "Update mode line."
  (interactive)
  (setq quilt-mode-line
	(concat " Q:" (quilt-short-patchname)))
  (force-mode-line-update))

(defun quilt-revert (filelist)
  "Refresh contents, editability and modeline of FILELIST.
FILELIST won't be touched unless their file is a child of the
current quilt directory.  Each elements in FILELIST should be the absolute
file names of those files affected by the latest quilt
operation.  Associated buffers get reverted to update their
contents.  Other buffers will only get their modeline and
editability adjusted."
  (let ((qd (quilt-dir))
	fn)
    (dolist (buf (buffer-list))
      (if (not (string-match "^ " (buffer-name buf)))
	(with-current-buffer buf
	  (setq fn (quilt-buffer-file-name-safe))
	  (when (string-equal qd (quilt-dir))
	    (quilt-update-modeline)
	    (when (and (not (buffer-modified-p))
		       (quilt-owned-p fn))
	      ;; If the file doesn't exist on disk it can't be reverted, but we
	      ;; need the revert hooks to run anyway so that the buffer's
	      ;; editability will update. Files not affected by the latest change
	      ;; (as listed in filelist) don't need to get reverted either.
	      (if (and (file-exists-p buffer-file-name)
		       (member fn filelist))
		  (progn
		    (revert-buffer t t t))
		(run-hooks 'after-revert-hook)))))))))

(defun quilt-push (arg)
  "Push next patch.  It is forced if ARG is specified."
  (interactive "P")
  (quilt-save)
  (if arg
      (quilt-cmd "push -f" "*quilt*")
    (quilt-cmd "push -q"))
  (quilt-revert (quilt-files-affected)))

(defun quilt-pop (arg)
  "Pop top patch.  It is forced if ARG is specified."
  (interactive "P")
  (quilt-save)
  (if arg
      (quilt-cmd "pop -f")
    (quilt-cmd "pop -q"))
  (quilt-revert (quilt-files-affected)))

(defun quilt-push-all (arg)
  "Push all remaining patches.  It is forced if ARG is specified."
  (interactive "P")
  (quilt-save)
  (let ((next (car (quilt-cmd-to-list "next"))))
    (if arg
	(quilt-cmd "push -f" "*quilt*")
      (quilt-cmd "push -qa"))
    (quilt-revert (quilt-files-affected next))))

(defun quilt-pop-all (arg)
  "Pop all applied patches.  It is forced if ARG is specified."
  (interactive "P")
  (quilt-save)
  (let ((fa (quilt-files-affected "-")))
    (if arg
	(quilt-cmd "pop -af")
      (quilt-cmd "pop -qa"))
    (quilt-revert fa)))

(defun quilt-goto ()
  "Go to a specified patch."
  (interactive)
  (let ((qd (quilt-dir)))
    (if (not qd)
	(quilt-cmd "push")		; to print error message
      (let ((arg (quilt-complete-list "Goto patch: " (quilt-patch-list))))
	(if (string-equal arg "")
	    (message "no patch name is supplied")
	  (quilt-save)
	  (let (cmd first last)
	    (if (file-exists-p (concat qd (quilt-pc-directory) "/" arg))
		(progn
		  (setq cmd "pop")
		  (setq first (car (quilt-cmd-to-list (concat "next " arg))))
		  (setq last (quilt-top-patch)))
	      (setq cmd "push")
	      (setq last arg)
	      (setq first (if (quilt-bottom-p)
			      "-"
			    (car (quilt-cmd-to-list "next")))))
	    (quilt-cmd (concat cmd " -q " arg) "*quilt*")
	  (quilt-revert (quilt-files-affected first last))))))))

(defun quilt-top ()
  "Display topmost patch."
  (interactive)
  (quilt-cmd "top"))

(defun quilt-find-file ()
  "Find a file in the topmost patch."
  (interactive)
  (let ((qd (quilt-dir)))
    (if (not qd)
	(quilt-cmd "push")		; to print error message
      (if (quilt-bottom-p)
	  (quilt-cmd "applied")		; to print error message
	(let ((l (quilt-file-list)))
	  (if (not l)
	      (message "no file is existed in this patch")
	    (let ((f (quilt-complete-list "File: " l)))
	      (if (string-equal f "")
		  (message "file name is not specified")
		(find-file (concat qd f))))))))))

(defun quilt-files ()
  "Display files in topmost patch."
  (interactive)
  (quilt-cmd "files"))

(defun quilt-import (fn pn)
  "Import external patch FN as PN.patch."
  (interactive "fPatch to import: \nsPatch name: ")
  (if (not pn)
      (message "no patch name supplied")
    (quilt-cmd (concat "import -p " pn ".patch " (if fn fn pn)))))

(defun quilt-diff ()
  "Display the diff of current change."
  (interactive)
  (quilt-save)
  (quilt-cmd "diff" "*diff*"))

(defun quilt-new (f)
  "Create a new patch F.patch."
  (interactive "sPatch name: ")
  (if (string-equal f "")
      (message "no patch name is supplied")
    (quilt-save)
    (quilt-cmd (concat "new " f ".patch"))
    (quilt-revert nil)))

(defun quilt-applied ()
  "Show applied patches."
  (interactive)
  (quilt-cmd "applied" "*quilt*"))

(defun quilt-add (arg)
  "Add ARG to the current patch."
  (interactive "b")
  (with-current-buffer arg
    (let ((fn (quilt-buffer-file-name-safe)))
      (cond
       ((not fn)
	(message "buffer %s is not associated with any buffer" (buffer-name)))
       ((not (quilt-dir))
	(quilt-cmd (concat "push")))	; to print error message
       (t
	(quilt-cmd (concat "add " (quilt-drop-dir fn)))
	(quilt-revert (list fn)))))))

(defun quilt-edit-patch ()
  "Edit the topmost patch."
  (interactive)
  (let ((qd (quilt-dir)))
    (if (not qd)
	(quilt-cmd "push")		; to print error message
      (let ((patch (quilt-top-patch)))
	(if (not patch)
	    (quilt-cmd "applied")	; to print error message
	  (quilt-save)
	  (let ((pf (concat qd
			    (format "/%s/" (quilt-patches-directory))
			    patch)))
	    (if (file-exists-p pf)
		(progn (find-file pf)
		       (read-only-mode 0))
	      (message (format "%s doesn't exist yet." pf)))))))))

(defun quilt-patches ()
  "Show which patches modify the current buffer."
  (interactive)
  (let ((fn (quilt-buffer-file-name-safe)))
    (cond
     ((not fn)
      (message "buffer %s is not associated with any buffer" (buffer-name)))
     ((not (quilt-dir))
      (quilt-cmd "push"))		; to print error message
     (t
      (quilt-cmd (concat "patches " (quilt-drop-dir fn)))))))

(defun quilt-unapplied ()
  "Display unapplied patch list."
  (interactive)
  (quilt-cmd "unapplied" "*quilt*"))

(defun quilt-refresh ()
  "Refresh the current patch."
  (interactive)
  (quilt-save)
  (quilt-cmd "refresh"))

(defun quilt-remove ()
  "Remove a file from the current patch and revert it."
  (interactive)
  (let ((f (quilt-buffer-file-name-safe)))
    (cond
     ((not f)
      (message "buffer %s is not associated with any patch" (buffer-name)))
     ((not (quilt-dir))
      (quilt-cmd "push")) ; to print error message
     (t
      (if (quilt-bottom-p)
	  (quilt-cmd "applied")		; to print error message
	(let ((dropped (quilt-drop-dir f)))
	  (if (y-or-n-p (format "Really drop %s? " dropped))
	      (progn
		(quilt-cmd (concat "remove " dropped))
		(quilt-revert (list f))))))))))

(defun quilt-edit-series ()
  "Edit the patch series file."
  (interactive)
  (let ((qd (quilt-dir)))
    (if (not qd)
	(quilt-cmd "push")		; to print error message
      (let ((series (concat qd
			     (format "/%s/series" (quilt-patches-directory)))))
	(if (file-exists-p series)
	    (find-file series)
	  (message (quilt-top-patch)))))))

(defun quilt-header (arg)
  "Print the header of ARG."
  (interactive "P")
  (let ((qd (quilt-dir)))
    (if (not qd)
	(quilt-cmd "push")		; to print error message
      (if (not arg)
	  (quilt-cmd "header")
	(let ((p (quilt-complete-list "Patch: " (quilt-patch-list))))
	  (if (string-equal p "")
	      (message "no patch name is supplied")
	    (quilt-cmd (concat "header " p))))))))

(defun quilt-delete (arg)
  "Delete ARG from the series file."
  (interactive "P")
  (let ((qd (quilt-dir)))
    (if (not qd)
	(quilt-cmd "push")		; to print error message
      (let ((p (if arg
		   (quilt-complete-list "Delete patch: " (quilt-patch-list))
		 (quilt-top-patch))))
	(if (string-equal p "")
	    (message "no patch name is supplied")
	  (if (not p)
	      (quilt-cmd "applied")	; to print error message
	    (if (y-or-n-p (format "Really delete %s? " p))
		(let ((fa (quilt-files-affected p p)))
		  (quilt-save)
		  (quilt-cmd (concat "delete " p))
		  (quilt-revert fa)))))))))

(defvar quilt-header-directory nil)

(defun quilt-header-commit ()
  "Commit to change patch header."
  (interactive)
  (let ((tmp (make-temp-file "quilt-header-")))
    (set-visited-file-name tmp)
    (basic-save-buffer)
    (cd quilt-header-directory)
    (shell-command (concat "EDITOR=cat quilt -r header <" tmp))
    (kill-buffer (current-buffer))
    (delete-file tmp)))

(defvar quilt-header-mode-map (make-keymap))
(define-key quilt-header-mode-map "\C-c\C-c" 'quilt-header-commit)

(defun quilt-edit-header (arg)
  "Edit the header of ARG."
  (interactive "P")
  (let ((qd (quilt-dir)))
    (if (not qd)
	(quilt-cmd "push")		; to print error message
      (let ((p (if arg
		   (quilt-complete-list "Edit patch: " (quilt-patch-list))
		 (quilt-top-patch))))
	  (if (string-equal p "")
	      (message "no patch name is supplied")
	    (if (not p)
		(quilt-cmd "applied")	; to print error message
	      (let ((qb (get-buffer-create (format " *quilt-header(%s)*" p))))
		(switch-to-buffer-other-window qb)
		(erase-buffer)
		(kill-all-local-variables)
		(make-local-variable 'quilt-header-directory)
		(setq quilt-header-directory default-directory)
		(use-local-map quilt-header-mode-map)
		(setq major-mode 'quilt-header-mode)
		(call-process "quilt" nil qb nil "header" p)
		(goto-char 0))))))))

(defun quilt-series (arg)
  "Show patche series.  It can be verbose mode if ARG is specified."
  (interactive "P")
  (if arg
      (quilt-cmd "series -v")
    (quilt-cmd "series")))

(defvar quilt-mode-map (make-sparse-keymap))
(define-key quilt-mode-map "\C-c.t" 'quilt-top)
(define-key quilt-mode-map "\C-c.f" 'quilt-find-file)
(define-key quilt-mode-map "\C-c.F" 'quilt-files)
(define-key quilt-mode-map "\C-c.d" 'quilt-diff)
(define-key quilt-mode-map "\C-c.p" 'quilt-push)
(define-key quilt-mode-map "\C-c.o" 'quilt-pop)
(define-key quilt-mode-map "\C-c.P" 'quilt-push-all)
(define-key quilt-mode-map "\C-c.O" 'quilt-pop-all)
(define-key quilt-mode-map "\C-c.g" 'quilt-goto)
(define-key quilt-mode-map "\C-c.A" 'quilt-applied)
(define-key quilt-mode-map "\C-c.n" 'quilt-new)
(define-key quilt-mode-map "\C-c.i" 'quilt-import)
(define-key quilt-mode-map "\C-c.a" 'quilt-add)
(define-key quilt-mode-map "\C-c.e" 'quilt-edit-patch)
(define-key quilt-mode-map "\C-c.m" 'quilt-patches)
(define-key quilt-mode-map "\C-c.u" 'quilt-unapplied)
(define-key quilt-mode-map "\C-c.r" 'quilt-refresh)
(define-key quilt-mode-map "\C-c.R" 'quilt-remove)
(define-key quilt-mode-map "\C-c.s" 'quilt-series)
(define-key quilt-mode-map "\C-c.S" 'quilt-edit-series)
(define-key quilt-mode-map "\C-c.h" 'quilt-header)
(define-key quilt-mode-map "\C-c.H" 'quilt-edit-header)
(define-key quilt-mode-map "\C-c.D" 'quilt-delete)

(defvar quilt-mode nil)
(make-variable-buffer-local 'quilt-mode)

(defun quilt-mode (&optional arg)
  "Toggle 'quilt-mode'.  Enable 'quilt-mode' if ARG is positive.

\\{quilt-mode-map}"
  (interactive "P")
  (setq quilt-mode
	(if (null arg)
	    (not quilt-mode)
	  (> (prefix-numeric-value arg) 0)))
  (if quilt-mode
      (let ((f (quilt-buffer-file-name-safe)))
	(if (quilt-owned-p f)
	    (if (not (quilt-editable f))
            (read-only-mode 1)
	      (read-only-mode 0)))
	(quilt-update-modeline))))

(defun quilt-hook ()
  "Enable quilt mode for quilt-controlled files."
  (if (quilt-p) (quilt-mode 1)))

(add-hook 'find-file-hooks 'quilt-hook)
(add-hook 'after-revert-hook 'quilt-hook)

(or (assq 'quilt-mode minor-mode-alist)
    (setq minor-mode-alist
	  (cons '(quilt-mode quilt-mode-line) minor-mode-alist)))

(or (assq 'quilt-mode-map minor-mode-map-alist)
    (setq minor-mode-map-alist
	  (cons (cons 'quilt-mode quilt-mode-map) minor-mode-map-alist)))

(provide 'quilt)

;;; quilt.el ends here
